### ISSUE TITLE
🔨 [Tracker] Move auto-attack damage out to lua

### ISSUE THREAD

**Issue by WinterSolstice8** [DerpyProjectGroup/topaz#4949]
The goal of this tracker is to show progress on moving out _all_ auto attack damage to lua (weaponskills will be handled later).
Many downstream users want to customize this -- rightfully so, there's many ways one may want to change things. Given both the complexity and frankly unreasonable amount of copypasted code to simply take damage this needs to happen at the same time as a refactor.

Until it's all fleshed out, melee auto attacks in lua will be disabled by default and there will be a setting to opt-in to test it

- [ ] Create unified damage taken function in lua that does _all_ the things -- PDT,MDT,DT, SDT, weakness to weapons, phalanx, stoneskin, magic stoneskin, magic phalanx, etc etc. Currently there's so many copypastes of how to take damage it's unreasonable to change how it works.
- [ ] fSTR/fSTR2
- [x] cRatio/wRatio etc
- [ ] weapon damage
- [ ] weapon rank
- [ ] multiattacks
- [ ] TP calculation
- [ ] dDEX
- [ ] dAGI
- [ ] add effects
- [ ] spikes interaction
- [x] blocking
- [x] parrying
- [x] guarding
- [ ] counter
- [ ] kicks
- [ ] daken
- [ ] more that I forgot...

**Comment by TeoTwawki** [comment]
when we get there, let me handle spikes and add effects. I plan a full refactor for spikes and when I redid add effects I made it really easy to migrate that. Only caveat Is I am planning on cleaning up the add effect global before starting on the spikes refactor.  It'll be awhile.

**Comment by zach2good** [comment]
I'll need to do some profiling, but I believe that the thing that costs cycles is going from C++ out into a Lua method call. I don't think calling usertype member functions is nearly as expensive, so we'd want this to be something like:
```cpp
LuaDamageResult result = lua["xi"]["damage"]["getDamage"](a, bunch, of, stuff);
auto* PChar = result.attacker;
auto* PMob = result.defender;
auto* action = result.action;
// etc. 
```
and that would be it. Inside Lua it can be as long as you'd possibly want, but it won't hop out into multiple C++->Lua function calls. 

* Again, I need to profile this to be absolutely sure, but this is how I think it plays out by intuition (I could be very wrong 🤷 )

**Comment by zach2good** [comment]
Made an issue to track this testing: https://github.com/LandSandBoat/server/issues/4951

**Comment by zach2good** [comment]
> Until it's all fleshed out, melee auto attacks in lua will be disabled by default and there will be a setting to opt-in to test it

Yes. I love you. You sane human being.

**Comment by Rbolomey** [comment]
So I just signed up for Github to learn more about landsandboat since I started playing on some ffxi private servers the last couple of years. 

   I was going to make a separate issue report until I saw this and thought you guys might stumble upon this yourselves. But it is driving me crazy. I am hoping that when you build out these separate luas you can please align both melee and WS to use the same pDIF equations. 

   Recently I noticed melee damage on the server I play can be derived back to pdifs on bgwiki capped per weapon type. Ref https://www.bg-wiki.com/index.php?title=PDIF&oldid=695037. 

   However, weaponskills are all capped at 2.625 pre-randomizer with no change per weapon type. I took a look at the physical utilities lua and the weaponskills lua and the equations governing capping c-ratio, level correction, and pdif are different than the ones in the physical utilities. The weaponskill lua seems to be referencing the old pre 2016 equations before SE updated them. Ref https://www.bg-wiki.com/index.php?title=PDIF&oldid=277998. The key indicator of this is the "cratio = utils.clamp(cratio, 0, 2.25)" on line 1195 in the weaponskill lua and no reference to weapontype caps anywhere.

I also think it would align more closely with Montenten's derived damage model if the w-ratio/c-ratio values per weapon type were capped at the effective max for the lower limit and max for the upper limit before going into the pdif function. 

If you think I should make a separate issues post for this I will do that please just let me know.

**Comment by WinterSolstice8** [comment]
yep they'll be using the same functions and include that

**Comment by Xaver-DaRed** [comment]
The physical_utilities.lua file is mostly unused for the time being. The long-term plan is to re-code damage calculations properly into physical_utilities and such files and use those. Similarly to how its being done with magic.

**Comment by Rbolomey** [comment]
Yea Xaver, I just saw your pull on cure spells using the new damage functions. This is exciting. I can't wait to see how things turn out^^. Great job guys.
